### DISCUSSION TITLE
Mixed mode debugging - is it possible?

### DISCUSSION THREAD

**Discussion by michaldobrodenka** [kitamstudios/rust-analyzer.vs#18] [Ideas]
I'm thinking about replacing some parts of my solution with rust. 

Do you think it can be possible to seamlessly debug when I step between c# and rust code compiled as native library?

**Comment by parthopdas** [comment]
yep that should work.

if you want to browse both C# and Rust you can open your entire code as a folder. C# works reasonably well. alternatively use 2 IDE - one for C# another for Rust.

let me know how it goes, i am interested in helping you get going.

**Reply by michaldobrodenka** [reply]
I tried one folder with rust library, compiled. Opened in one VS2022 + rust-analyzer.vs extension. In second VS2022 I had c# solution, with one csproj project, calling rust like this:

```
        [DllImport("c:\\work\\test\\rust\\cs_call_rst\\target\\debug\\our_rust.dll")]
        private static extern Int32 add_numbers(Int32 number1, Int32 number2);
```

When I'm debugging c#, step in doesn't go into rust code. So I tried to break VS2022 in c# code, attach VS2022 with opened rust sources - rust VS2022 can't attach because it's saying that this process has attached debugger already.
So I've put `Console.ReadLine()` into c# source to wait, run it without debug, while waiting for Enter attached VS2022 with rust sources. It is attached but breakpoints are empty circles and won't be hit.

**Reply by parthopdas** [reply]
hmm - this should work - our_rust.dll is just another native dll. please can you check if the symbols have been loaded for that?

alternatively - if you can share a repro project i can help make it work. pretty sure it is something trivial.

**Reply by michaldobrodenka** [reply]
> hmm - this should work - our_rust.dll is just another native dll. please can you check if the symbols have been loaded for that?
> 
> alternatively - if you can share a repro project i can help make it work. pretty sure it is something trivial.

[call_rust_dll.zip](https://github.com/kitamstudios/rust-analyzer.vs/files/11136328/call_rust_dll.zip)

Hi, I'm attaching folder with 2 projects: 

call_rust - csproj & solution C# - to call rust dll
cs_call_rst - folder with rust lib 

I open each of the project in separate VS2022. Maybe I should add rust folder to c# sln somehow? but I don't know how. Thank you!

**Reply by parthopdas** [reply]
apologies for the delay - i will get to this by tomorrow.

**Reply by parthopdas** [reply]
wth 5 min job why keep it for tomorrow.

you need to enable native code debugging as well. since it is a c# project by default only managed code debugging is enabled. [here is how to do it](https://learn.microsoft.com/en-us/visualstudio/debugger/how-to-debug-managed-and-native-code?view=vs-2022)

this is what i see:

![image](https://user-images.githubusercontent.com/2140460/230474908-563ce4f2-3857-4249-a21a-e004661901e5.png)

**Reply by michaldobrodenka** [reply]
woooow! I must try that! Thank you!

**Reply by michaldobrodenka** [reply]
It works! Thank you. Next step for me would be mixed mode debugging on android net7-android to rust dll :)

**Reply by parthopdas** [reply]
okay. please keep me posted on the overall architecture of your split work, issues you face & what tools can help.

@ericsink's https://github.com/ericsink/native-aot-samples can help imho.

ps: much gratitude for the sponsorship 💖

**Reply by Jashwanth537** [reply]
how do i give cargo commands such as cargo run here
